apis/Google.Cloud.Logging.Console/Google.Cloud.Logging.Console.Tests/GoogleCloudConsoleFormatterTest.cs

```diff
@@ -31,13 +35,108 @@ public void SimplestLog()
             var expectedJson = "{\"message\":\"test\",\"severity\":\"INFO\"}\n";
             var actualJson = writer.ToString();
 
-            Assert.Equal(actualJson, expectedJson);
+            Assert.Equal(expectedJson, actualJson);
+        }
+
+        [Fact]
+        public void InvalidLogLevel_Throws()
+        {
+            var formatter = CreateFormatter();
+            var logEntry = new LogEntry<string>((LogLevel)8, "LogCategory", new EventId(1), "test", exception: null, (state, exception) => state);
+            var writer = new StringWriter { NewLine = "\n" };
+
+            Assert.Throws<ArgumentOutOfRangeException>(() => formatter.Write(logEntry, scopeProvider: null, writer));
+        }
+
+        [Fact]
+        public void NoMessageAndNoException_Nothing()
+        {
+            var formatter = CreateFormatter();
+            var logEntry = new LogEntry<string>((LogLevel)8, "LogCategory", new EventId(1), null, exception: null, (state, exception) => state);
+            var writer = new StringWriter { NewLine = "\n" };
+            formatter.Write(logEntry, scopeProvider: null, writer);
+
+            var expectedJson = "";
+            var actualJson = writer.ToString();
+
+            Assert.Equal(expectedJson, actualJson);
+        }
+
+        [Fact]
+        public void Log_CorrectScopeInformation()
+        {
+            var formatter = CreateFormatter(new GoogleCloudConsoleFormatterOptions() { IncludeScopes = true });
+
+            var logEntry = new LogEntry<string>(LogLevel.Information, "LogCategory", new EventId(1), "test", exception: null, (state, exception) => state);
+            var writer = new StringWriter { NewLine = "\n" };
+
+            var scopeProvider = new LoggerExternalScopeProvider();
+
+            formatter.Write(logEntry, scopeProvider: scopeProvider, writer);
+            var expectedJson = "{\"message\":\"test\",\"severity\":\"INFO\",\"scopes\":[]}\n";
+            var actualJson = writer.ToString();
+
+            Assert.Equal(expectedJson, actualJson);
+        }
+
+
+        [Fact]
+        public void Log_KeyValuePair()
+        {
+            var formatter = CreateFormatter();
+
+            var _state = new ReadOnlyCollection<KeyValuePair<string, object>>(new List<KeyValuePair<string, object>>()
+            {
+                new KeyValuePair<string, object>("A","Sample value 1"),
+                new KeyValuePair<string, object>("B","Sample value 2"),
+            });
+
+            var logEntry = new LogEntry<IReadOnlyCollection<KeyValuePair<string, object>>>(LogLevel.Information, "LogCategory", new EventId(1), _state, exception: null, (state, exception) => state.ToString());
+            var writer = new StringWriter { NewLine = "\n" };
+
+            formatter.Write(logEntry, scopeProvider: null, writer);
+            
+            var actualJson = writer.ToString();
+
+            Assert.Contains("Sample value 1", actualJson);
```

Comment:
I suspect there's more we can assert here. I suspect it may be worth parsing the JSON and checking it more structurally than textually.

Reply:
Aggree.

Reply:
like using dynamic?
```
dynamic actualState = JsonConvert.DeserializeObject(actualJson);

            Assert.Contains("Sample value 1", actualJson);
            Assert.Contains("Sample value 2", actualJson);
            Assert.NotNull(actualState.state.A);
            Assert.NotNull(actualState.state.B);
            Assert.Equal("Sample value 1", actualState.state.A.Value);
            Assert.Equal("Sample value 2", actualState.state.B.Value);
```

Reply:
No, I wouldn't use dynamic typing. I'd parse it to a `JObject` or the equivalent in System.Text.Json.

Reply:
is this ok?
```
var actualJson = writer.ToString();

var actualState = JsonDocument.Parse(actualJson);
var actualValueA = actualState.RootElement.GetProperty("state").GetProperty("A").GetString();
var actualValueB = actualState.RootElement.GetProperty("state").GetProperty("B").GetString();

Assert.Contains("Sample value 1", actualJson);
Assert.Contains("Sample value 2", actualJson);
Assert.Equal("Sample value 1", actualValueA);
Assert.Equal("Sample value 2", actualValueB);
```

Reply:
Yes, I think that would be fine.

Reply:
Sorry, I focused on the bits I *did* want to see before - I hadn't spotted that you'd kept the Assert.Contains part. There's no need for that given the more specific tests for property A and property B.